jwst/ramp_fitting/ramp_fit_step.py

```diff
@@ -242,11 +246,43 @@ def process(self, input):
                 self.save_opt, readnoise_2d, gain_2d, self.algorithm,
                 self.weighting, max_cores, dqflags.pixel, suppress_one_group=self.suppress_one_group)
 
-        # Save the OLS optional fit product, if it exists
-        if opt_info is not None:
-            opt_model = create_optional_results_model(input_model, opt_info)
-            self.save_model(opt_model, 'fitopt', output_file=self.opt_name)
+            # To calculate the readnoise variance for weighting, ramp_fitting
+            # will be run a second time - this time with the DO_NOT_USE flags
+            # removed from the UNDERSAMP-flagged pixels. To reduce execution time,
+            # fitting will effectively be performed only on the UNDERSAMP-flagged
+            # pixels, by flagging all other pixels as DO_NOT_USE.
+
+            # First create an intermediate model ('_W' for weighting) to
+            # input to a 2nd ramp_fitting call.
+            input_model_W = copy.copy(input_model)
+
+            # Flag all pixels as DO_NOT_USE, then remove the DO_NOT_USE flag
+            # from the UNDERSAMP-flagged pixels.
+            gdq = input_model_W.groupdq.copy()
+            wh_undersamp = np.where(np.bitwise_and(gdq.astype(np.int32), dqflags.group['UNDERSAMP']))
+            gdq = np.bitwise_or(gdq, dqflags.group['DO_NOT_USE'])
+            gdq[wh_undersamp] -= dqflags.group['DO_NOT_USE']
+            input_model_W.groupdq = gdq
+
+            # Rerun fitting to get the weighted RN variances
+            image_info_W, integ_info_W, opt_info_W, gls_opt_model_W = ramp_fit.ramp_fit(
+                input_model_W, buffsize,
+                self.save_opt, readnoise_2d, gain_2d, self.algorithm,
+                self.weighting, max_cores, dqflags.pixel, suppress_one_group=self.suppress_one_group)
 
+            # The VAR_RNOISE arrays are written out to the out_model via the image_info tuple,
+            # and similarly for the integration and optional models. Create new info tuples
+            # to be composed of the original values plus the weighted RN variance value. The
+            # 3rd component of each contains the RN variance.
+            image_info_new = (image_info[0], image_info[1], image_info[2], image_info_W[3], image_info[4])
```

Comment:
This will fail if `image_info` is NoneType.  Checking that non-NoneType values are returned from `ramp_fit` should occur before trying to access elements of the returns.

Reply:
Updated that and related code.